### ISSUE TITLE
Prover gets confused about \fresh if many objects are allocated (static checking)

### ISSUE THREAD

**Issue by slyubomirsky** [OpenJML/OpenJML#458]
So far, in cases where I've used methods that ensure their result is a freshly allocated object (via ``ensures \fresh(\result)``), the prover has generally been able to successfully prove that two objects allocated by that method are not reference-equal. However, in cases where I've allocated many objects by that method, the prover is no longer able to conclude that two objects allocated by the method are distinct.

I've constructed a test case below where I reproduce this error:
Allocator.java:
```
package alloc;
public class Allocator {
	public static byte[] alloc(int size) {
		return new byte[size];
	}
}
```

Allocator.jml:
```
package alloc;
public class Allocator {
	
	/*@ public normal_behavior
	  @   requires size >= 0 && size < 128;
	  @   assignable \nothing;
	  @   ensures \fresh(\result);
	  @   ensures \result.length == size;
	  @   ensures \forall int i; 0 <= i && i < size; \result[i] == (byte) 0x00;
	  @*/
	public static byte[] alloc(int size);
}
```

The following test passes without problems:
```
	public void testAllocSmall() {
		byte[] b1 = Allocator.alloc(10);
		//@ assert b1 != null;
		//@ assert b1.length == 10;
		
		byte[] b2 = Allocator.alloc(10);
		//@ assert b2 != null;
		//@ assert b2.length == 10;
		//@ assert b1 != b2;
	}
```

However, if I allocate more objects between b1 and b2, suddenly, the prover fails to establish that ``b1 != b2``. I've attached the full test case below; here is an abridged version:

```
		byte[] b1 = Allocator.alloc(10);
		//@ assert b1 != null;
		//@ assert b1.length == 10;
		
		byte[] b2 = Allocator.alloc(11);
		//@ assert b2 != null;
		//@ assert b2.length == 11;
		//@ assert b2 != b1;
		
                [...]

		byte[] b9 = Allocator.alloc(18);
		//@ assert b9 != null;
		//@ assert b9.length == 18;
		//@ assert b1 != b9;
		//@ assert b2 != b9;
		//@ assert b3 != b9;
		//@ assert b4 != b9;
		//@ assert b5 != b9;
		//@ assert b6 != b9;
		//@ assert b7 != b9;
		//@ assert b8 != b9;

		byte[] b10 = Allocator.alloc(10);
		//@ assert b10 != b1;
}
```

The last assert fails, even though per the \fresh() assertion, it should not. There are no other errors when I run static checking (with Z3) on this case. (Notice that the failure only occurs with objects produced by the method with the same arguments -- the prover has no trouble telling apart b1 and b9 when they were produced with different arguments.)

[FreshTest.zip](https://github.com/OpenJML/OpenJML/files/394725/FreshTest.zip)

**Comment by davidcok** [comment]
I've been (and am) on holiday but will be looking at this tomorrow.
- David

**Comment by slyubomirsky** [comment]
Can you give any word on where in the OpenJML codebase there could be a problem, if you know where? I've run into further problems relating to freshness/distinctness of objects and my colleagues and I are willing to investigate and make potential changes to OpenJML (with a pull request) if we can get some information as to what might need to change

**Comment by davidcok** [comment]
Steven - I'm on this. Here is one bit of information. In your longer example, you allocated different lengths of arrays in each allocate call. Hence each new allocation had a different length, and OpenJML could determine that each new allocation was distinct because of the different lengths. It is more challenging if you make all the allocations the same length - and then the problem you point out needs only three allocations. If you want to look into this and similar problems, try adding the option -show. You will then see the logical transformations in detail and can inspect them for encoding errors.
- David

**Comment by davidcok** [comment]
Fixed in V 0.7.0, now released
